riscv/processor.cc

```diff
@@ -107,6 +107,27 @@ static bool check_pow2(int val)
   return ((val & (val - 1))) == 0;
 }
 
+void processor_t::set_pmp_num(reg_t pmp_num) {
+
+  // check the number of pmp is in a reasonable range
+  if( pmp_num >= 0 && pmp_num <= 16) {
+    state.platform_n_pmp = pmp_num;
+  } else {
+      fprintf(stderr, "error: bad number of pmp regions: '%ld' from the dtb\n", pmp_num);
+      abort();
+  }
+}
+
+void processor_t::set_pmp_granularity(reg_t pmp_granularity) {
+  // check the pmp granularity is set from dtb(!=0) and is power of 2
+  if (pmp_granularity >= 4 && ((pmp_granularity & (pmp_granularity - 1)) == 0)) {
+    state.pmp_granularity = pmp_granularity;
```

Comment:
Since you are using `state.pmp_granularity` as a shift amount elsewhere, don't you need a `ctz` here?

Reply:
the minimum granularity is when G=0 in 2^(G+2).

Reply:
No, but I mean, `pmp_granularity` is `2^(G+2)`, whereas elsewhere in the code you are using `state.pmp_granularity` to mean `G+2`, so you need to do `state.pmp_granularity = ctz(pmp_granularity)` to take the logarithm of the RHS.